hail/src/main/scala/is/hail/expr/ir/Emit.scala

```diff
@@ -2430,6 +2432,71 @@ private class Emit(
           }
         }
 
+      case x@NDArraySlice(child, slicesIR) =>
+        val childEmitter = deforest(child)
+
+        val slicest = emit(slicesIR, env, resultRegion, None)
+        val slicesValueAddress = mb.newField[Long]
+        val slicesTuple = new CodePTuple(coerce[PTuple](slicesIR.pType), region, slicesValueAddress)
+
+        val codeSlices = ArrayBuffer[(Code[Long], Code[Long], Code[Long])]()
+        val codeRefMap = mutable.Map[Int, Code[Long]]()
+
+        var missingSliceElements = const(false)
+
+        coerce[PTuple](slicesIR.pType).types.zipWithIndex.foreach { case (sliceOrIndex, dim) =>
+          sliceOrIndex match {
+            case p: PTuple => {
+              val oneSlice = new CodePTuple(p, region, slicesTuple(dim))
+              missingSliceElements = missingSliceElements || oneSlice.isMissing(0) || oneSlice.isMissing(1) || oneSlice.isMissing(2)
+              codeSlices += ((oneSlice[Long](0), oneSlice[Long](1), oneSlice[Long](2)))
+            }
+            case _: PInt64 => {
+              codeRefMap += dim -> slicesTuple(dim)
+            }
+          }
+        }
+
+        val outputShape = codeSlices.toArray.map { case (start, stop, step) =>
+          (step >= 0L && start <= stop).mux(
+            const(1L) + ((stop - start) - const(1L)) / step,
+            (step < 0L && stop <= start).mux(
+              (((stop - start) + const(1L)) / step) + const(1L),
+              0L)
+          )
+        }
+
+        val setupMissing = Code(
+          slicest.setup,
+          slicesValueAddress := slicest.value[Long],
+          childEmitter.setupMissing
+        )
+
+        val setupShape = Code(
+          childEmitter.setupShape
+        )
+
+        val missing = childEmitter.missing || missingSliceElements
+
+        new NDArrayEmitter(mb, x.pType.nDims, outputShape, x.pType.shape.pType, x.pType.elementType, setupShape, setupMissing, missing) {
+          override def outputElement(idxVars: Array[Code[Long]]): Code[_] = {
+            val oldIdxVarsIter = idxVars.iterator
+            val sliceIdxVarsIter = codeSlices.iterator
+
+            val sliceIdxVars = Array.tabulate(childEmitter.nDims) { dim =>
+              if (codeRefMap.contains(dim)) {
+                codeRefMap(dim)
+              } else {
+                val (start, _, step) = sliceIdxVarsIter.next()
+                val oldIdxVar = oldIdxVarsIter.next()
+                start + oldIdxVar * step
+              }
+            }
```

Comment:
I think this bit would be more clear if transformed the wider bit to look more
like this:
```scala
val slicers = slicesTuple.withTypes.collect {
    case (t: PTuple, slice) => new CodePTuple(t, region, slice) }
val anyMissingness = slicesTuple.missingness.fold(_ || _) ||
    slicers.map(_.missingness.fold(_ || _)).fold(_ || _)
val shape = slicers.map(_.value[Long, Long, Long]).map { case (start, stop, step) =>
  ...
}.toArray

// ...

new ...
  override def outputElement ....
    val inIdx = idxVars.iterator
    val sliceIdxVars = slicesTuple.withTypes.map {
      case (_: PInt64, indexer, i) =>
        coerce[Long](indexer)
      case (t: PTuple, slice) =>
        val (start, _, step) = val CodePTuple(t, region, slice).value[Long, Long, Long]
        start + inIdx.next * step
    }
    childEmitter.outputElement(sliceIdxVars)
```
I think this is conceptually a bit easier to follow